lib/metrics.js

```diff
@@ -22,169 +14,68 @@ class MetricsEvent {
-    const deduplicateStream = new DeduplicateStream()
-    const splitStream = new PassThrough({ objectMode: true })
-    deduplicateStream
-      .pipe(new TelemetryStream(moduleName))
-      .pipe(splitStream)
-    splitStream
-      .pipe(new SerializeStream())
-      .pipe(
-        this.logs.createWriteStream(
-          join(this.metricsPath, `${moduleName}.log`)
-        )
-      )
-    splitStream
-      .pipe(new MergeMetricsStream(...await Promise.all([
-        this.getLatest(),
-        this.getLatest(moduleName)
-      ])))
-      .pipe(new SerializeStream())
-      .pipe(this.logs.createWriteStream(this.allMetricsPath))
-    return deduplicateStream
-  }
-
-  async maybeCreateMetricsFile (moduleName) {
-    await maybeCreateFile(
-      this.#getMetricsFilePath(moduleName),
-      formatLog(
-        JSON.stringify({ totalJobsCompleted: 0, totalEarnings: '0' }) + '\n'
-      )
-    )
-  }
-}
-
-class DeduplicateStream extends Transform {
-  constructor () {
-    super({ objectMode: true })
-    this.last = null
-  }
-
-  _transform (obj, _, callback) {
-    const isChanged = this.last === null ||
-        Object
-          .entries(obj)
-          .some(([key, value]) => this.last[key] !== value)
-    if (isChanged) {
-      this.last = obj
-      callback(null, obj)
-    } else {
-      callback()
-    }
-  }
-}
-
-class TelemetryStream extends Transform {
-  constructor (moduleName) {
-    assert(moduleName, 'moduleName required')
-    super({ objectMode: true })
-    this.lastMetrics = null
-    this.moduleName = moduleName
-  }
-
-  _transform (metrics, _, callback) {
-    if (typeof this.lastMetrics?.totalJobsCompleted === 'number') {
+  submit (moduleName, metrics) {
+    if (
+      typeof this.moduleMetrics.get(moduleName)?.totalJobsCompleted === 'number'
+    ) {
       writeClient.writePoint(
         new Point('jobs-completed')
-          .stringField('module', this.moduleName)
+          .stringField('module', moduleName)
           .intField(
             'value',
-            metrics.totalJobsCompleted - this.lastMetrics.totalJobsCompleted
+            metrics.totalJobsCompleted -
+              this.moduleMetrics.get(moduleName).totalJobsCompleted
           )
       )
     }
-    this.lastMetrics = metrics
-    callback(null, metrics)
-  }
-}
-
-let mergedMetrics
-
-class MergeMetricsStream extends Transform {
-  constructor (allMetrics, moduleMetrics) {
-    assert.strictEqual(
-      typeof allMetrics?.totalJobsCompleted,
-      'number',
-      'allMetrics.totalJobsCompleted required'
-    )
-    assert.strictEqual(
-      typeof moduleMetrics?.totalJobsCompleted,
-      'number',
-      'moduleMetrics.totalJobsCompleted required'
-    )
-    super({ objectMode: true })
-    if (!mergedMetrics) {
-      mergedMetrics = allMetrics
+    this.moduleMetrics.set(moduleName, metrics)
+    const mergedMetrics = Object
+      .values(this.moduleMetrics)
+      .reduce((acc, metrics) => {
```

Comment:
❗ 
I think you have a bug in this code, which may also explain why I am not seeing any metric updates.

The first hint was offered by TypeScript, which was inferring both `acc` and `metrics` as `any`.

Apparently, you cannot use `Object.values()` on a map, you need to use `Map.prototype.values()` instead.

```
❯ node
Welcome to Node.js v18.14.2.
Type ".help" for more information.
> m = new Map();
Map(0) {}
> m.set("foo", "bar")
Map(1) { 'foo' => 'bar' }
> m
Map(1) { 'foo' => 'bar' }
> Object.values(m)
[]
> m.values()
[Map Iterator] { 'bar' }
```

Unfortunately, `[Map Iterator]` does not provide `map` method. An easy solution is to convert the iterator into a new array, paying the cost of extra allocation & copying:

```
> Array.from(m.values()).map(x => console.log(x))
bar
```

Personally, I would ditch `reduce` and replace it with a for-of loop:

```js
> for (x of m.values()) { console.log(x); }
bar
```

Reply:
Great catch! Fixed using a for loop as suggested, it's also shorter this way